library/core/src/num/mod.rs

```diff
@@ -87,27 +87,103 @@ depending on the target pointer size.
     };
 }
 
+macro_rules! widening_impl {
+    ($SelfT:ty, $WideT:ty, $BITS:literal) => {
+        /// Calculates the full multiplication of `self` and `rhs`.
+        ///
+        /// This returns the low-order (insignificant) bits and the high-order (significant) bits
+        /// of the result as two separate values, in that order.
+        ///
+        /// # Examples
+        ///
+        /// Basic usage:
+        ///
+        /// Please note that this example is shared between integer types.
+        /// Which explains why `u32` is used here.
+        ///
+        /// ```
+        /// #![feature(widening_mul)]
+        /// assert_eq!(5u32.widening_mul(2), (10, 0));
+        /// assert_eq!(1_000_000_000u32.widening_mul(10), (1410065408, 2));
+        /// ```
+        #[unstable(feature = "widening_mul", issue = "none")]
+        #[rustc_const_unstable(feature = "const_widening_mul", issue = "none")]
+        #[must_use = "this returns the result of the operation, \
+                      without modifying the original"]
+        #[inline(always)]
+        pub const fn widening_mul(self, rhs: Self) -> (Self, Self) {
+            // note: longer-term this should be done via an intrinsic,
+            //   but for now we can deal without an impl for u128/i128
+            // SAFETY: overflow will be contained within the wider types
+            let wide = unsafe { (self as $WideT).unchecked_mul(rhs as $WideT) };
+            (wide as $SelfT, (wide >> $BITS) as $SelfT)
+        }
```

Comment:
I believe `widening_mul` on signed integers should return the less-significant half as unsigned instead. For example consider `i8`, and −128 × 3 (`0x80 * 0x03`). Now, when widened to `i16` (`0xff80 * 0x0003`), this gives −384 (`0xfe80`), which is then returned as (`0x80`, `0xfe`). Now, `0xfe` itself is −2, so think of this as meaning −512 because it is the higher word. The correct result is −512 + 128, so the `0x80` should be interpreted as +128 not −128, so it would be `u8`, not `i8`.

Note: I'm using unsigned for all hex values to not make it more confusing, even though technically `0xfe_i8` is an error and would be written as `-0x02_i8`.

Reply:
Also, using unsigned for the lower half would make it possible to write something like `real_result = (i16::from(high) << 8) | i16::from(low)`, or `+` instead of `|`, both work. With signed for the lower half, `i16::from(low)` would be sign extended and would actually need to be masked with something like `(i16::from(low) & i16::from(u8::MAX))`.

Reply:
I should add, these kinds of suggestions are exactly the kind that I would love to see discussed, and I would definitely suggest bringing it up in the original RFC for `widening_mul`. Right now I'm kind of inclined to keep this particular method as-is until it's merged and this discussion can happen before stabilisation, since the main goal is to get something out there for people to use at the moment.

I am fine with making this change before stabilisation though if anyone else would like to +1 the idea

Reply:
Also FWIW, my justification for using the same type is because any big-integer implementation that uses these would be adding one half or the other to some part of the bigint anyway, and using a separate unsigned type for that would just require unnecessary casting. I could easily be swayed, though!

Reply:
I have looked into the internal representation of two bigint implementations: GMP and num-bigint. Both store the digits as unsigned and use an extra flag to indicate whether the big number is positive or negative, so I don't know how an actual bignum implementation that stores digits as signed would use these, if there is such a bignum implementation.

Reply:
Yeah, I did notice that most bigint implementations do use a sign flag instead of two's complement. However, in my case, I was working on an implementation of a variable-length integer encoding where integers would still be stored in two's complement as it would make sign extension easier. It probably does make sense for most bigint implementations to use a sign flag and unsigned int units, but that doesn't mean that two's complement bigints are completely out of the question.

Reply:
About the ordering of the return tuple: I guess it makes sense to be consistent with other methods who's carry is the last element of the return tuple, and as such have the most significant word on the right. Personally, I prefer the most significant word on the left; then one can write nice idiomatic code e.g. to compare `x * y` with `a * b` with ```x.widening_mul(y).cmp(&a.widening_mul(b))```.

Reply:
I definitely think that `add_with_carry` matching the order of `overflowing_add` is the most important part.

But interestingly, the least-significant word on the left has a nice interpretation on little-endian machines: the return is just the number for `wide_mul` (assuming the obvious, though unspecified, tuple layout): https://rust.godbolt.org/z/sG9E4G8vY

(Though apparently we no longer encode `(u32, u32)` as `i64` in LLVM, which is why the Godbolt demo uses an old rustc version, so maybe that matters less these days.)